lib/std/mem.zig

```diff
@@ -635,13 +635,79 @@ test "lessThan" {
 /// Compares two slices and returns whether they are equal.
 pub fn eql(comptime T: type, a: []const T, b: []const T) bool {
     if (a.len != b.len) return false;
-    if (a.ptr == b.ptr) return true;
+    if (a.len == 0 or a.ptr == b.ptr) return true;
+    if (@sizeOf(T) == 0) return true;
+
+    // No vectorization in stage2 x86_64 or x86
+    // https://github.com/ziglang/zig/issues/17748
+    if (builtin.zig_backend != .stage2_x86_64 and builtin.zig_backend != .stage2_x86) {
+        if (std.meta.hasUniqueRepresentation(T)) return eqlBytes(std.mem.sliceAsBytes(a), std.mem.sliceAsBytes(b));
+    }
+
     for (a, b) |a_elem, b_elem| {
         if (a_elem != b_elem) return false;
     }
     return true;
 }
 
+/// std.mem.eql heavily optimized for slices of bytes.
+pub fn eqlBytes(a: []const u8, b: []const u8) bool {
+    if (a.len != b.len) return false;
+    if (a.len == 0 or a.ptr == b.ptr) return true;
+
+    if (a.len <= 16) {
+        if (a.len < 4) {
+            const x = (a[0] ^ b[0]) | (a[a.len - 1] ^ b[a.len - 1]) | (a[a.len / 2] ^ b[a.len / 2]);
+            return x == 0;
+        }
+        var x: u32 = 0;
+        for ([_]usize{ 0, a.len - 4, (a.len / 8) * 4, a.len - 4 - ((a.len / 8) * 4) }) |n| {
+            x |= @as(u32, @bitCast(a[n..][0..4].*)) ^ @as(u32, @bitCast(b[n..][0..4].*));
+        }
+        return x == 0;
+    }
+
+    // Figure out the fastest way to scan through the input in chunks.
+    // Uses vectors when supported and falls back to usize/words when not.
+    const Scan = if (std.simd.suggestVectorSize(u8)) |vec_size|
+        struct {
+            pub const size = vec_size;
+            pub const Chunk = @Vector(size, u8);
+            pub inline fn isZero(chunk: Chunk) bool {
+                return @reduce(.Or, chunk) == 0;
+            }
+        }
+    else
+        struct {
+            pub const size = @sizeOf(usize);
+            pub const Chunk = usize;
+            pub inline fn isZero(chunk: Chunk) bool {
+                return chunk == 0;
+            }
+        };
+
+    inline for (1..6) |s| {
+        const n = 16 << s;
+        if (n <= Scan.size and a.len <= n) {
+            const V = @Vector(n / 2, u8);
+            var x = @as(V, a[0 .. n / 2].*) ^ @as(V, b[0 .. n / 2].*);
+            x |= @as(V, a[a.len - n / 2 ..][0 .. n / 2].*) ^ @as(V, b[a.len - n / 2 ..][0 .. n / 2].*);
+            return @reduce(.Or, x) == 0;
+        }
+    }
+    // Compare inputs in chunks at a time (excluding the last chunk).
+    for (0..(a.len - 1) / Scan.size) |i| {
+        const a_chunk: Scan.Chunk = @bitCast(a[i * Scan.size ..][0..Scan.size].*);
+        const b_chunk: Scan.Chunk = @bitCast(b[i * Scan.size ..][0..Scan.size].*);
+        if (!Scan.isZero(a_chunk ^ b_chunk)) return false;
```

Comment:
Is there a reason why you do
```zig
!(@reduce(.Or, a_chunk ^ b_chunk) == 0)
```
instead of
```zig
@reduce(.Or, a_chunk != b_chunk)
```
here?

Looking at [godbolt](https://godbolt.org/z/857nnfqEM), both versions produce the exact same code in release.
However the second version does seem better at conveying the intent here and it also seems to produce less code in debug mode, since the boolean Or reduction can be done in a single instruction on x86, whereas the bitwise Or reduction does take log(n) operations.

Reply:
Huh, interesting. There is no particular reason for it other than I didn't think it would affect it, but looking at the godbolt I tend to agree with you. It seems the reduce is emitting much cleaner code because of the better defined intent as you say. I have no problem changing it :)

Reply:
After a couple more performances tests to compare them, I see absolutely no difference between the two methods. Neither in ReleaseFast nor in Debug. I do not see the point of changing it now to be honest. The CI has run for a while, and it would be unfortunate to re-run it for no actual gain. To be honest, the code seems more ungainly now, then before. `isZero` just makes more sense when reading the flow.g

I will continue testing tomorrow, and if I find something that increases performance, I'll be sure to add it :smile:.

Reply:
So, as it turns out, I was given a reason to switch to this bool method 😄 
stage2_x86_64 does not currently support non-bool reduces. Changing it to only use bool reduces keeps the same exact performance, while allowing stage2 to make use of it.

Thank you for the idea!